### ISSUE TITLE
Make type tagged intrinsic work in generic context

### ISSUE THREAD

**Issue by tindzk** [scala-native/scala-native#396]
`sizeof` correctly resolves the implicit `ClassTag` parameter:

```scala
println(42 * sizeof[Int])
```

This should be equivalent to: 

```scala
def size[T](length: Int)(implicit ct: ClassTag[T]): CSize = length * sizeof[T]
println(size[Int](42))
```

However, the compiler crashes with the following exception:

```
scala.scalanative.util.package$UnsupportedException: ct (class scala.reflect.internal.Trees$Ident)
```

**Comment by densh** [comment]
Thanks for the bug report!

**Comment by densh** [comment]
Current API that rely on tags that are affected by this:

1. Indirect function calls through function pointers.
2. C-style varargs.
3. Stack and zone allocation. 
4. Sizeof.
5. Reinterpret casts. 
6. Pointer operations.

**Comment by jokade** [comment]
I've encountered a similar problem. I get 
```
java.util.NoSuchElementException: None.get
	at scala.None$.get(Option.scala:347)
	at scala.None$.get(Option.scala:345)
	at scala.scalanative.nscplugin.NirGenUtil$class.unwrapTagOption(NirGenUtil.scala:74)
```
during NIR generation for
```scala
class NSArray[T <: NSObject] protected (implicit evidence$1: ClassTag[T]) extends NSObject with NSCopying with NSMutableCopying with NSSecureCoding {
  // ...
  @inline def arrayByAddingObject(anObject: T): NSArray[T] = _root_.objc.helper.objc_msgSend1[T](this, NSArray.__sel_arrayByAddingObject_, anObject).cast[Object].asInstanceOf[NSArray[T]];
}
```
where `_root_.objc.helper.objc_msgSend1` is a def macro:
```scala
 @inline def objc_msgSend1[T1](self: id, op: SEL, arg1: T1): id = macro Macros.objc_msgSend1Impl[T1]

class Macros(val c: Context) {
  import c.universe._

  val tPtrByte = weakTypeOf[scalanative.native.Ptr[Byte]]

  def objc_msgSend1Impl[T1: c.WeakTypeTag](self: c.Tree, op: c.Tree, arg1: c.Tree): c.Expr[id] =
    c.Expr[id](q"_root_.objc.helper.objc_msgSendPtr.cast[scalanative.native.CFunctionPtr3[$tPtrByte,$tPtrByte,${c.weakTypeOf[T1]},$tPtrByte]].apply($self,$op,$arg1)")
}
```

**To reproduce this error**:
- requires scala-native with PR #1132 
- check out https://github.com/jokade/scalanative-cocoa/commit/325337ab321593f89c8ea7dcf80ef5fb51dba398
- `sbt foundation/compile`

**Comment by densh** [comment]
Fixed in #1426
